### PR TITLE
Add RNN PTB to official models

### PR DESCRIPTION
**Pull request by k-w-w**
Hi Asim, can you review the PTB model? I followed the object-oriented style, tests and benchmarks from the MNIST model. 

Thanks!

### CODE REVIEW

File: official/rnn_ptb/run_training.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+# number steps to train the model).
+TRAIN_FILE_WORDS = 929589
+
+
+def input_fn(input_file, vocab_dict, num_epochs=1):
+  """Generates batches of data from the input file."""
+  def _input_fn(params):
+    batch_size, unrolled_count = params.batch_size, params.unrolled_count
+
+    def batch_data(data, steps_per_epoch):
```
Comment (karmel) [inline official/rnn_ptb/run_training.py]: Nesting functions three levels deep seems to introduce unnecessary complexity; could just be my misunderstanding, but any reason not to extract these and have them take the args they need? Is there an advantage to having them defined in the _input_fn?

File: official/rnn_ptb/run_training.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        steps_per_epoch: The number of steps it take to complete an epoch.
+      """
+      data = data.reshape([batch_size, steps_per_epoch, unrolled_count])
+      return np.swapaxes(data, 0, 1)
+
+    def input_generator():
+      """Yield batches of the input data."""
+      # Read all words from the input file, and append <eos> at every new line.
+      with tf.gfile.GFile(input_file, 'r') as f:
+        data = f.read().replace('\n', '<eos>').split()
```
Comment (karmel) [inline official/rnn_ptb/run_training.py]: Do we anticipate users trying files/input with non-unix line endings? Might be worth at least extracting these to module-level vars (something like `INPUT_EOL`, `EOL_TOKEN` or what-have-you), just to make it easier for eventual users to edit. (I speak from the experience of having tried to adapt a similar model in the past...)

Edit: I see below that EOS is defined in `util`; should that be used here as well?

File: official/rnn_ptb/util.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+def build_vocab_id_dict(data_file):
+  """Create a dictionary mapping all words in the vocabulary to integer IDs."""
+  assert tf.gfile.Exists(data_file), 'File %s does not exist.' % data_file
+  with tf.gfile.Open(data_file, 'r') as f:
+    unique_words = set(f.read().split())
+
+    # Remove special symbols from the set.
+    for symbol in SPECIAL_SYMBOLS:
+      if symbol in unique_words:
+        unique_words.remove(symbol)
```
Comment (karmel) [inline official/rnn_ptb/util.py]: nit: I think this loop is equivalent to `unique_words = unique_words - set(SPECIAL_SYMBOLS)`

File: official/rnn_ptb/util.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    # is mapped to 0 (corresponding to '<unk>').
+    vocab_dict = collections.defaultdict(
+        int, zip(unique_words, range(len(unique_words))))
+    return vocab_dict
+
+
+def build_reverse_vocab_dict(vocab_dict):
+  """Build a dictionary of id->word."""
+  reverse_dict = {i: word for word, i in vocab_dict.iteritems()}
+  reverse_dict[3] = '.\n'  # Replace '<eos>' with a period with a new line.
```
Comment (karmel) [inline official/rnn_ptb/util.py]: Why append a period here? Seems like this could lead to unexpected behavior if we necessarily add punctuation. Also, hard-coding `3` here might be fragile if someone adds/changes the SPECIAL_SYMBOLS, which seems like a reasonable thing to do. Maybe at least do a quick lookup of the index of EOS in SPECIAL_SYMBOLS?

File: official/rnn_ptb/README.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+2. **Train the model with [`run_training.py`](run_training.py):**
+   ```
+   python run_training.py --data_dir={DATA PATH}
+   ```
+   Required arguments:
+   * `--data_dir`: path to the training/validation/testing files. (e.g. `.../Downloads/simple-examples/data`)
+
+   Other arguments:
+   * `--model`: The model hyperparameters/configuration to use.
+         Options are: small, `medium` (default), and `large`.
```
[the remark this answers is not in the document]
Comment (k-w-w) [reply official/rnn_ptb/README.md]: Thanks for catching this! 

What do you mean by example/format?

File: official/rnn_ptb/README.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+   * `--model_dir`: directory where the model will be saved. The model will be saved to `/tmp/ptb_model` by default.
```
[the remark this answers is not in the document]
Comment (k-w-w) [reply official/rnn_ptb/README.md]: Done, I left `_model` part in (`/tmp/rnn_ptb_model`), because the other models have `_model` as part of the directory name.

File: official/rnn_ptb/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (karmel) [reply official/rnn_ptb/README.md]: Ah, wait, I think I misread initially and thought you could pass in arbitrary configs. Disregard!

File: official/rnn_ptb/run_training.py (hunk 1 above)
Comment (k-w-w) [reply official/rnn_ptb/run_training.py]: I originally had them under `_input_fn` for convenience. I agree that 3 nested functions is a bit too much, so I've moved them out.

File: official/rnn_ptb/run_training.py (hunk 2 above)
Comment (k-w-w) [reply official/rnn_ptb/run_training.py]: Good point, I added a constant for `\n`, and reuse util.EOS for `<eos>`.

File: official/rnn_ptb/util.py (hunk 3 above)
Comment (k-w-w) [reply official/rnn_ptb/util.py]: Nice! I've made this change.

File: official/rnn_ptb/util.py (hunk 4 above)
Comment (k-w-w) [reply official/rnn_ptb/util.py]: I've changed this to replace `<eos>` with the NEWLINE symbol, and added a lookup for the index of EOS. I agree it's better with less hardcoding.

File: official/rnn_ptb/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (k-w-w) [reply official/rnn_ptb/README.md]: Ok!

**Comment by tfboyd** [comment]
This is about a year old at this point. Closing.  Ping me to reopen if desired.  Cleaning up all PRs targeting /official.
